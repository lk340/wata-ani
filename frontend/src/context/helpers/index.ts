import { createUseContext } from "./create-use-context";

export { createUseContext };
