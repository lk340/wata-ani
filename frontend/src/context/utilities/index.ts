import * as Theme from "./_theme";
import * as Window from "./_window";
import * as WindowScroll from "./_scroll";
import * as Mouse from "./_mouse";
import * as Location from "./_location";

export { Theme, Window, WindowScroll, Mouse, Location };
