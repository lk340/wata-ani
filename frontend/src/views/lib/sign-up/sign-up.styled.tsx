import styled, { css, createGlobalStyle } from "styled-components";
import { animated } from "react-spring";

import * as Snippets from "@/utils/style/snippets";

export const SignUp = styled("div")`
	${Snippets.fillContainer()};

	border: red solid 1px;
`;

export const SignUpForm = styled("form")`
	${Snippets.flex("column")};

	border: red solid 1px;
`;

export const SignUpFormTitle = styled("h1")`
	${Snippets.clear()};
	text-align: center;
`;

export const SignUpFormError = styled("div")`
	color: red;
`;

export const SignUpFormUsernameInput = styled("input").attrs(() => ({
	placeholder: "Username",
	required: true,
}))`
	outline: none;
`;

export const SignUpFormEmailInput = styled(SignUpFormUsernameInput).attrs(() => ({
	placeholder: "Email",
	type: "email",
	required: true,
}))``;

export const SignUpFormPasswordInput = styled(SignUpFormUsernameInput).attrs(() => ({
	placeholder: "Password",
	type: "password",
	required: true,
}))``;

export const SignUpFormPasswordConfirmationInput = styled(SignUpFormUsernameInput).attrs(
	() => ({
		placeholder: "Confirm Password",
		type: "password",
		required: true,
	}),
)``;

export const SignUpFormSubmit = styled("button").attrs(() => ({
	type: "submit",
}))``;
