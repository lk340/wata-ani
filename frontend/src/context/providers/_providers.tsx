import * as React from "react";

import * as Context from "@/context";

export const Providers: React.FC<{ children: React.ReactNode }> = ({ children }) => {
	return (
		<>
			<Context.Theme.Provider>
				<Context.Window.Provider>
					<Context.Scroll.Provider>
						<Context.Mouse.Provider>
							<Context.Location.Provider>
								<Context.Auth.Provider>
									<Context.UserAgent.Provider>
										<Context.Welcome.Provider>{children}</Context.Welcome.Provider>
									</Context.UserAgent.Provider>
								</Context.Auth.Provider>
							</Context.Location.Provider>
						</Context.Mouse.Provider>
					</Context.Scroll.Provider>
				</Context.Window.Provider>
			</Context.Theme.Provider>
		</>
	);
};
