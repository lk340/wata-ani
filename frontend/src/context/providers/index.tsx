import * as React from "react";

import * as Context from "@/context";

export const Providers: React.FC<{ children: React.ReactNode }> = ({ children }) => {
	return (
		<>
			<Context.Theme.Provider>
				<Context.Windows.Provider>
					<Context.Scroll.Provider>
						<Context.Mouse.Provider>
							<Context.Location.Provider>
								<Context.UserAgent.Provider>
									<Context.AuthForm.Provider>
										<Context.Navbar.Provider>
											<Context.Pagination.Provider>
												<Context.NavbarOptionsCreate.Provider>
													<Context.NavbarOptionsSettings.Provider>
														{children}
													</Context.NavbarOptionsSettings.Provider>
												</Context.NavbarOptionsCreate.Provider>
											</Context.Pagination.Provider>
										</Context.Navbar.Provider>
									</Context.AuthForm.Provider>
								</Context.UserAgent.Provider>
							</Context.Location.Provider>
						</Context.Mouse.Provider>
					</Context.Scroll.Provider>
				</Context.Windows.Provider>
			</Context.Theme.Provider>
		</>
	);
};

// const providers = [
// 	Context.Theme.Provider,
// 	Context.Window.Provider,
// 	Context.Scroll.Provider,
// 	Context.Mouse.Provider,
// 	Context.Location.Provider,
// 	Context.Auth.Provider,
// 	Context.UserAgent.Provider,
// 	Context.Filter.Provider,
// 	Context.Sort.Provider,
// 	Context.Demo.Provider,
// 	Context.Home.Provider,
// ];

// function generateProvider(providers: React.ReactNode[]) {
// 	if (providers.length === 1) return providers[0];

// 	const lastIndex = providers.length - 1;

// }

// const Children: React.FC<{ children: React.ReactNode }> = ({ children }) => {
// 	return <>{children}</>;
// };
