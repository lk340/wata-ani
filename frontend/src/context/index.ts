// ================= //
// ↓↓↓ Utilities ↓↓↓ //
// ================= //

import * as UserAgent from "@/context/utilities/user-agent";
import * as Location from "@/context/utilities/location";
import * as Windows from "@/context/utilities/windows";
import * as Scroll from "@/context/utilities/scroll";
import * as Theme from "@/context/utilities/theme";
import * as Mouse from "@/context/utilities/mouse";

// ============ //
// ↓↓↓ Data ↓↓↓ //
// ============ //

import * as AuthForm from "@/context/data/auth-form";
import * as Navbar from "@/context/data/navbar";
import * as Pagination from "@/context/data/pagination";
import * as NavbarOptionsCreate from "@/context/data/navbar-options-create";
import * as NavbarOptionsSettings from "@/context/data/navbar-options-settings";

// =============== //
// ↓↓↓ Exports ↓↓↓ //
// =============== //

export {
	UserAgent,
	Location,
	Windows,
	Scroll,
	Theme,
	Mouse,
	AuthForm,
	Navbar,
	Pagination,
	NavbarOptionsCreate,
	NavbarOptionsSettings,
};
