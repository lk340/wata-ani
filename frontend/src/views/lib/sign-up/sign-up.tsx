import * as React from "react";

import * as Context from "@/context";

import * as Styled from "./sign-up.styled";

export const SignUp = () => {
	const { auth } = Context.Auth.useAuthContext();

	React.useEffect(() => {
		auth.state.username !== "" &&
		auth.state.email !== "" &&
		auth.state.password !== "" &&
		auth.state.passwordConfirmation !== ""
			? auth.setters.setDisabled(false)
			: auth.setters.setDisabled(true);
	}, [
		auth.state.username,
		auth.state.email,
		auth.state.password,
		auth.state.passwordConfirmation,
	]);

	return (
		<Styled.SignUp>
			<Styled.SignUpForm onSubmit={auth.handlers.handleSubmit}>
				<Styled.SignUpFormTitle>Sign Up</Styled.SignUpFormTitle>
				<Styled.SignUpFormError>{auth.state.error}</Styled.SignUpFormError>
				<Styled.SignUpFormUsernameInput onChange={auth.handlers.handleUsernameChange} />
				<Styled.SignUpFormEmailInput onChange={auth.handlers.handleEmailChange} />
				<Styled.SignUpFormPasswordInput onChange={auth.handlers.handlePasswordChange} />
				<Styled.SignUpFormPasswordConfirmationInput
					onChange={auth.handlers.handlePasswordConfirmationChange}
				/>
				<Styled.SignUpFormSubmit disabled={auth.state.disabled}>
					Sign Up
				</Styled.SignUpFormSubmit>
			</Styled.SignUpForm>
		</Styled.SignUp>
	);
};
