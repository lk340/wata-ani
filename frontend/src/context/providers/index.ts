import { Providers } from "./_providers";
export { Providers };
