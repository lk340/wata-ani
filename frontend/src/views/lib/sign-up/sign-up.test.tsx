import * as React from "react";
import * as Testing from "@testing-library/react";
import axiosMock from "axios";

import { Providers } from "@/context/providers";

import { SignUp } from "./sign-up";

jest.mock("axios");

function renderWithContext(component: React.ReactNode) {
	return Testing.render(<Providers>{component}</Providers>);
}

describe("SignUp", () => {
	it("matches the snapshot", () => {
		const { asFragment } = renderWithContext(<SignUp />);
		expect(asFragment(<SignUp />)).toMatchSnapshot();
	});

	it(`displays the title with the text "Sign Up"`, () => {
		const { getByTestId } = renderWithContext(<SignUp />);
		expect(getByTestId("title")).toHaveTextContent("Sign Up");
	});

	it("displays username, email, password, and password confirmation input fields", () => {
		const { getByTestId } = renderWithContext(<SignUp />);
		expect(getByTestId("username-field")).not.toBeNull();
		expect(getByTestId("email-field")).not.toBeNull();
		expect(getByTestId("password-field")).not.toBeNull();
		expect(getByTestId("password-confirmation-field")).not.toBeNull();
	});

	it("displays a submit button", () => {
		const { getByTestId } = renderWithContext(<SignUp />);
		expect(getByTestId("submit-button")).not.toBeNull();
	});

	it("disables the submit button when all form fields are not filled", () => {
		const { getByTestId } = renderWithContext(<SignUp />);
		const usernameField = getByTestId("username-field");
		const emailField = getByTestId("email-field");
		const passwordField = getByTestId("password-field");
		Testing.fireEvent.change(usernameField, { target: { value: "foobar" } });
		Testing.fireEvent.change(emailField, { target: { value: "foo@bar.com" } });
		Testing.fireEvent.change(passwordField, { target: { value: "password123" } });
		expect(getByTestId("username-field").value).toBe("foobar");
		expect(getByTestId("email-field").value).toBe("foo@bar.com");
		expect(getByTestId("password-field").value).toBe("password123");
		expect(getByTestId("password-confirmation-field").value).toBe("");
		expect(getByTestId("submit-button")).toBeDisabled();
	});

	it("enables the submit button when all form fields are filled", () => {
		const { getByTestId } = renderWithContext(<SignUp />);
		const usernameField = getByTestId("username-field");
		const emailField = getByTestId("email-field");
		const passwordField = getByTestId("password-field");
		const passwordConfirmationField = getByTestId("password-confirmation-field");
		Testing.fireEvent.change(usernameField, { target: { value: "foobar" } });
		Testing.fireEvent.change(emailField, { target: { value: "foo@bar.com" } });
		Testing.fireEvent.change(passwordField, { target: { value: "password123" } });
		Testing.fireEvent.change(passwordConfirmationField, {
			target: { value: "password123" },
		});
		expect(getByTestId("username-field").value).toBe("foobar");
		expect(getByTestId("email-field").value).toBe("foo@bar.com");
		expect(getByTestId("password-field").value).toBe("password123");
		expect(getByTestId("password-confirmation-field").value).toBe("password123");
		expect(getByTestId("submit-button")).not.toBeDisabled();
	});

	it("displays an error message when the passwords don't match", () => {
		const { getByTestId } = renderWithContext(<SignUp />);
		const usernameField = getByTestId("username-field");
		const emailField = getByTestId("email-field");
		const passwordField = getByTestId("password-field");
		const passwordConfirmationField = getByTestId("password-confirmation-field");
		Testing.fireEvent.change(usernameField, { target: { value: "foobar" } });
		Testing.fireEvent.change(emailField, { target: { value: "foo@bar.com" } });
		Testing.fireEvent.change(passwordField, { target: { value: "foo" } });
		Testing.fireEvent.change(passwordConfirmationField, { target: { value: "bar" } });
		Testing.fireEvent.click(getByTestId("submit-button"));
		expect(getByTestId("username-field").value).toBe("foobar");
		expect(getByTestId("email-field").value).toBe("foo@bar.com");
		expect(getByTestId("password-field").value).toBe("foo");
		expect(getByTestId("password-confirmation-field").value).toBe("bar");
		expect(getByTestId("error-message")).toHaveTextContent("Your passwords don't match.");
	});

	it("signs the user up", async () => {
		const { getByTestId } = renderWithContext(<SignUp />);
		const username = "foobar";
		const email = "foo@bar.com";
		const password = "password123";
		const passwordConfirmation = "password123";
		axiosMock.post.mockResolvedValueOnce({
			data: { id: 1, username, email },
		});
		const usernameField = getByTestId("username-field");
		const emailField = getByTestId("email-field");
		const passwordField = getByTestId("password-field");
		const passwordConfirmationField = getByTestId("password-confirmation-field");
		Testing.fireEvent.change(usernameField, { target: { value: username } });
		Testing.fireEvent.change(emailField, { target: { value: email } });
		Testing.fireEvent.change(passwordField, { target: { value: password } });
		Testing.fireEvent.change(passwordConfirmationField, {
			target: { value: passwordConfirmation },
		});
		Testing.fireEvent.click(getByTestId("submit-button"));
		expect(getByTestId("username-field").value).toBe(username);
		expect(getByTestId("email-field").value).toBe(email);
		expect(getByTestId("password-field").value).toBe(password);
		expect(getByTestId("password-confirmation-field").value).toBe(passwordConfirmation);
		expect(getByTestId("submit-button")).not.toBeDisabled();
		expect(axiosMock.post).toHaveBeenCalledTimes(1);
	});
});
