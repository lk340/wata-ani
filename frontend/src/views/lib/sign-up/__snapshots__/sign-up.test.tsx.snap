// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp matches the snapshot 1`] = `
<DocumentFragment>
  .c0 {
  width: 100%;
  height: 100%;
  border: red solid 1px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: auto;
  -webkit-justify-content: auto;
  -ms-flex-pack: auto;
  justify-content: auto;
  -webkit-align-items: auto;
  -webkit-box-align: auto;
  -ms-flex-align: auto;
  align-items: auto;
  border: red solid 1px;
}

.c2 {
  margin: 0;
  padding: 0;
  text-align: center;
}

.c3 {
  color: red;
}

.c4 {
  outline: none;
}

.c5 {
  outline: none;
}

.c6 {
  outline: none;
}

.c7 {
  outline: none;
}

<div
    class="c0"
  >
    <form
      class="c1"
    >
      <h1
        class="c2"
        data-testid="title"
      >
        Sign Up
      </h1>
      <div
        class="c3"
        data-testid="error-message"
      />
      <input
        class="c4"
        data-testid="username-field"
        placeholder="Username"
        required=""
      />
      <input
        class="c5"
        data-testid="email-field"
        placeholder="Email"
        required=""
        type="email"
      />
      <input
        class="c6"
        data-testid="password-field"
        placeholder="Password"
        required=""
        type="password"
      />
      <input
        class="c7"
        data-testid="password-confirmation-field"
        placeholder="Confirm Password"
        required=""
        type="password"
      />
      <button
        class=""
        data-testid="submit-button"
        disabled=""
        type="submit"
      >
        Sign Up
      </button>
    </form>
  </div>
</DocumentFragment>
`;
